Update R chapter (#367)
NLeSC/guide: language_guides/r.md

R is a functional programming language and software environment for statistical computing and graphics: https://www.r-project.org/.

## Philosophy and typical use cases
R is particularly popular in the social, health, and biological sciences where it is used for statistical modeling. R can also be used for signal processing (e.g. FFT), machine learning, image analyses, and natural language processing. The R syntax is similar to that of Matlab and Python in terms of compactness and readability, which makes it a good prototyping language for science.

One of the strengths of R is the large number of available open source statistical packages, often developed by domain experts. For example, R-package [Seewave](http://rug.mnhn.fr/seewave/) is specialised in sound analyses. Packages are typically released on CRAN [The Comprehensive R Archive Network](http://cran.r-project.org).

## Some crucial differences with Python

> Are you familiar with Python? Then you'll find these remarks useful:

Author: @bouweandela I'd say **yes**, but this will imply a significant removal of @maltelueken's work. I'd like to know his opinion first.

> Are you familiar with Python? Then you'll find these remarks useful:

Reviewer: I only took over the chapter last year, so the great majority of the chapter was written before that. For me, replacing sections with links to other resources is fine.

> Are you familiar with Python? Then you'll find these remarks useful:

Author: ```suggestion
Are you familiar with Python? Then kickstart your R journey by reading this [blog post](https://towardsdatascience.com/the-starter-guide-for-transitioning-your-python-projects-to-r-8de4122b04ad).
```

> * R does not need a notebook to program interactively. Instead we recommend to use [RStudio](https://posit.co/products/open-source/rstudio/), an IDE that is installed separately.

Reviewer 2: And maybe recommending R studio should be left for the IDE section

* Numbering in R starts with 1 and not with 0.

> * When `x` is a list, `x[1]` is another list containing the first element, and `x[[1]]` is its **value**.

Reviewer 2: Is there a link to slicing in R?

Author: Slicing is the first of second episode of virtually any R tutorial out there. I prefer to stick to the main differences with Python.

# Coding style conventions
It is good to follow the R style conventions as [posted](http://adv-r.had.co.nz/Style.html) by Hadley Wickham, which is seems compatible with the R style convention as posted by [Google](https://google.github.io/styleguide/Rguide.xml).

## The `<-` operator

### TL;DR
Assigning variables with `<-` instead of `=` is recommended, although **most** of the time both are equivalent.

### Details

> One point in both style conventions that has resulted in some discussion is the '<-' syntax for variable assignment. In the majority of R tutorials and books you will see that authors use this syntax, e.g. 'a <- 3' to assign value 3 to object 'a'. Please note that R syntax 'a = 3' will preform exactly the same operation in 99.9% of situations. The = syntax has less keystrokes and could therefore be considered more efficient and readable. Further, the = syntax avoids the risk for typos like a < -1, which will produce a boolean if 'a' exists, and a <- 1 which will produce an object 'a' with a numeric value. Further, the = syntax may be more natural for those who already use it in other computing languages.

Author: This whole discussion about `<-` vs `=` seems a bit too long the scope fo this guide. Shall we consider reducing it to a link for those really interested? What do you think, @maltelueken?

> # Coding style conventions

Reviewer 3: Instead of writing our own style guide, it would be better to link to existing resources. E.g. how about mentioning [styler](https://github.com/r-lib/styler) + [lintr](https://github.com/r-lib/lintr)? Or are there better alternatives?

Reviewer: I agree that we should mention those two packages. There is also `formatR`, but I am not sure how stable it is at the moment.

We could refer to the [tidyverse style guide](https://style.tidyverse.org/) as one option?

Author: It would be good to clearly choose one. Do we have such a preference?

Author: We'll go for the tidyverse style. @PabRod will rewrite this section.

It is good to follow the R style conventions as [posted](http://adv-r.had.co.nz/Style.html) by Hadley Wickham, which is seems compatible with the R style convention as posted by [Google](https://google.github.io/styleguide/Rguide.xml).

> ## The `<-` operator

Reviewer 2: For the scope of the guide, I thing we should just give a recommendation "we recommend using `->`" (or something like that ;-) ).

The discussion on why use different operators is interesting, but outside of the scope of the guide. Maybe we can provide a pointer to an existing discussion "For a heated debate on `=` vs `<-` read here...". If we can point to an existing one, great. Otherwise maybe we can write a blog post with such a debate and link to it from here :-)

Author: A blog about this sounds like a great idea. What do you think, @maltelueken?

Author: There are already man blogs about this. @PabRod will take care of replacing the paragraph with a link to one of those.

- `addone(x=3)`
  - will produce `4` as it will assign `3` to known function argument `x`

The `<-` supporters will argue that this example demonstrates that `=` should be avoided. However, it also demonstrates that `=` syntax can work in the context of function input if `=` is only used for assigning values to input arguments that are expected by the function (`x` in the example above) and to never introduce new R objects as part of a function call (`b` in the example above).

From a computer science perspective it is probably best to adhere to the `<-` convention. From a domain science perspective it is understandable to use `=`. The code performs exactly the same and guarantees that new objects created as part of a function call result in an error. Please note that it is also possible to develop code with `=` syntax and to transfer it to `<-` syntax once the code is finished: the `formatR` package offers tools for doing this. The CRAN repository for R packages accepts both forms of syntax.

## `%>%` and `|>`
The symbols `%>%` and `|>` represent the pipe operator.

> The first one is part of the `magrittr` package, and it gained so much popularity that it was added as part of native R since version 4.0.0. For the rest, they are entirely equivalent.

Reviewer: ```suggestion
The first one is part of the `magrittr` package, and it gained so much popularity that a similar operator, `|>`, was added as part of native R since version 4.1.0. For details on the differences between the two, see this [blog post](https://www.tidyverse.org/blog/2023/04/base-vs-magrittr-pipe/).
```

From a computer science perspective it is probably best to adhere to the <- convention. From a domain science perspective it is understandable to use =. The code performs exactly the same and guarantees that new objects created as part of a function call result in an error. Please note that it is also possible to develop code with = syntax and to transfer it to <- syntax once the code is finished, the formatR package offers tools for doing this. The CRAN repository for R packages accepts both forms of syntax.

> ## `%>%` and `|>`

Reviewer: Perhaps add here that `|>` requires R > 4.0.0 which makes code incompatible with some legacy code. Also, `%>%` is not a native R operator and requires the `magrittr` package as a dependency (which is loaded with almost all tidyverse packages). Lean packages might not want to rely on this dependency.

Author: Done!

## Speeding up code
As in many computing languages loops should be avoided in R. Here is a list of tricks to speed up your code:

> * `read.table()` is sometimes faster than `read.csv()`

Author: Fixed in https://github.com/NLeSC/guide/pull/367/commits/22fa014ef1a7d465f54874c827987ab0cf965a17.

* `ifelse()`
* `lapply()`
* `sapply()`
* `mapply()`
* `grep()`
* `%in%` for testing whether and where values in one object occur in another object
* `aggregate()`
* `which()` for identifying which object indices match a certain condition
* `table()` for getting a frequency table of categorical data
* `grep()`
* `gsub()`
* [`dplyr`](http://dplyr.tidyverse.org/) package

> * [`purrr`](https://purrr.tidyverse.org/) package. It contains many functional programming tools that can help you build optimized processes

Reviewer: ```suggestion
* [`purrr`](https://purrr.tidyverse.org/) package. It contains many functional programming tools that can help you build optimized processes

If you need R code that is even more performant, consider implementing functions in C++ and calling them from R via the [`Rcpp`](https://www.rcpp.org/) package.
```

Customary R uses `.Rd` files in `/man` directory for documentation. These files and folders are automatically created by RStudio when you create a new project from your existing R-function files.

Function level comments starting with `#'` are used by `roxygen` to automatically generate the `.Rd` files. This means that you **don't have to edit the `.Rd` files directly**.

R function documentation offers plenty of space to document the functionality, including code examples, literature references, and links to related functions. Nevertheless, it can sometimes be helpful for the user to also have a more generic description of the package with for example use-cases. You can do this with a `vignette`.

Read more about vignettes in [Package documentation](http://r-pkgs.had.co.nz/vignettes.html) chapter of Hadleys [R packages](http://r-pkgs.had.co.nz) book.
Read more about `roxygen` syntax on it's [github page](https://github.com/yihui/roxygen2). `roxygen` will also populate `NAMESPACE` file which is necessary to manage package level imports.

# Available templates

> Most of the templating is nativelly managed by the [`usethis`](https://usethis.r-lib.org/) package. Said this, these others can serve as inspiration:

Reviewer 2: ```suggestion
Most of the templating is nativelly managed by the [`usethis`](https://usethis.r-lib.org/) package. Having said this, these others can serve as inspiration:
```

Reviewer 2: Also, maybe we should add (or point to) a snippet of code of how we recommend to use `usethis` as suggested in https://github.com/NLeSC/guide/issues/327
